Updated plugin detail view

jklausa opened the pull request:
Updates the view used for plugin details.

A fair share of hairs I've lost writing this of this was fighting with the WPOrg completely free-form HTML payloads for fields like `description` and `changelog`. 

I'm not... super proud of the choices made here, but I think they're reasonable-ish considering the constraints we have to work within.

To test: open the app, go to the plugin list of a Jetpack site and tap one, opening the new detail view.

koke commented on WordPress/Classes/Utility/WPImmuTableRows.swift:
```diff
@@ -107,15 +107,15 @@ struct CheckmarkRow: ImmuTableRow {
 
 struct LinkRow: ImmuTableRow {
-    static let cell = ImmuTableCell.class(WPTableViewCellValue1.self)
+    static let cell = ImmuTableCell.class(WPExternalLinkCell.self)
```
I'm not sure you need a custom subclass. Unless I missed some details, you could use a `WPTableViewCellDefault` and create/set the `accessoryView` in `configureCell(_:)`

jklausa replied:
You're right! I think I initially wrote it when I thought I'll add `ExternalLinkRow` in addition to `LinkRow`, but it wasn't needed in the first place at all.

koke commented:
If you toggle Slow Animations in the simulator, you can see that the chevron in the expandable cells does some odd things when animating. I'm not sure if the text is animating as desired.

![expandable-cell](https://user-images.githubusercontent.com/8739/35105217-56df5dac-fc6b-11e7-8702-b0d631fe8336.gif)

koke commented on WordPress/Classes/ViewRelated/Plugins/PluginViewModel.swift:
```diff
@@ -88,29 +105,114 @@ class PluginViewModel: Observable {
+            settingsLink = LinkRow(
+                title: NSLocalizedString("Settings", comment: "Link to plugin's Settings"),
+                action: { [unowned self] _ in
+                    self.presentBrowser(for: settingsURL)
```
The settings link will likely end up in wp-login, since we don't have credentials to pre-authenticate Jetpack users. Maybe it's better not to show it until we can log them in automatically? @elibud thoughts?

elibud replied:
> until we can log them in automatically?
☝️ will we ever be able to do that on wp-admin?

jklausa replied:
Hm — I think it's still worthwhile, since they can just tap to open it in Safari, where they could either be logged in or have the credentials stored in iCloud Keychain/1Password/etc. I know I do, but I guess it's probable that not many people do.

elibud replied:
I can't find that in Calypso, am I missing something or is it not there?

jklausa replied:
@elibud here's Akismet as an example:
<img width="722" alt="sshot" src="https://user-images.githubusercontent.com/1594081/35113084-915f2db2-fc80-11e7-8797-d6fbf3ffce9a.png">

elibud replied:
Thanks, just tried that, we'll be replicating Calypso behavior and I'm fine with that. My uneducated guess is that any user that actually wants to modify the plugin settings and has good reason for it would have access or know her credentials.

koke commented on WordPress/Classes/ViewRelated/Plugins/PluginViewModel.swift:
```diff
@@ -29,30 +29,47 @@ class PluginViewModel: Observable {
-                action: { [unowned self] (_) in
-                    ActionDispatcher.dispatch(PluginAction.update(id: self.plugin.id, site: self.site))
-                }
+        switch (plugin.state.version, plugin.state.updateState) {
```
Maybe we can extract `versionRow` to a property and just `guard let version = plugin.state.version else { return nil }` to keep the switch simpler?

jklausa replied:
Done!

koke commented on WordPress/Classes/ViewRelated/Plugins/PluginDetailViewHeaderCell.swift:
```diff
@@ -0,0 +1,67 @@
+
+        nameLabel?.text = directoryEntry.name
+
+        let author = directoryEntry.author!
```
Why is that `!` there? If there's no chance an entry has a missing author, the model should not be optional, if it can be nil, it should not crash

jklausa replied:
True, it was leftover when I was experimenting with stuff that I missed when cleaning up. Fixed!

koke commented on WordPress/Classes/ViewRelated/Plugins/PluginViewModel.swift:
```diff
@@ -29,30 +29,47 @@ class PluginViewModel: Observable {
+                    title: message,
+                    subtitle: subtitle,
+                    actionLabel: NSLocalizedString("Update", comment: "Button label to update a plugin"),
+                    action: { [unowned self] (_) in
```
I think this is the same behavior that before, but now that the "Update" looks like a button and not just a label on the cell, it feels odd that tapping anywhere else in the row will trigger an update.

jklausa replied:
Agree, changed the behavior.

jklausa commented:
Fixed up the animation — it's still not exactly perfect, but at least the chevron isn't doing crazy things. I've added it to my todo list for the polishing phase.

jklausa commented:
rebased the branch on `develop` to fix the merge conflicts.

koke commented:
> rebased the branch on develop to fix the merge conflicts

I've got a conflict on `git pull`. I think it's best to avoid rebasing branches that are already pushed and just use merge.

koke commented on WordPress/Classes/Utility/WPImmuTableRows.swift:
```diff
@@ -167,22 +180,48 @@ struct DestructiveButtonRow: ImmuTableRow {
-                action(self)
-            }
-        })
+        cell.onButtonPressed = { self.onButtonTap(self) }
```
Minor nitpick, but it'd be nicer if these two variables had the same name

koke commented:
The animation looks so much better now. The chevron still stretches a bit when going from collapsed to expanded, and the title jumps and moves when going from expanded to collapsed. I'm fine with polishing that on a future PR

koke commented:
I'm not sure about this, but tapping anywhere in the expandable cell will collapse, which might be problematic if you're trying to tap on a link and miss by a few pixels. I'd love to get some design input on that one

koke commented:
There are several details around font sizes in the new expandable cells

1. By default, the title font is slightly smaller than the other rows

![img_0430](https://user-images.githubusercontent.com/8739/35143464-eb89370c-fd01-11e7-9fcf-ec4e46986581.png)

2. When you turn on dynamic text, the titles don't take that into account

![img_0427](https://user-images.githubusercontent.com/8739/35143493-0c2f7a02-fd02-11e7-900f-3004dec32566.png)

3. The cell contents respect dynamic text settings, but only on first load. The rest of the cells change when you come back to the app from settings, but the expandable cells only reflect changes if you go back to the list and load the plugin detail again.

![img_0428](https://user-images.githubusercontent.com/8739/35143784-da81e0b6-fd02-11e7-8c98-339f6078d756.png)

4. The expanded cell contents seems to have an even smaller font than the title

![img_0429](https://user-images.githubusercontent.com/8739/35143777-d4bb5a4a-fd02-11e7-8e02-5952e85ca632.png)

jklausa commented:
Fixed issues 1, 2, 3. 

4 is how it looks like by design: 

<img width="565" alt="untitled" src="https://user-images.githubusercontent.com/1594081/35288685-b4f3a0a8-0065-11e8-8bca-9ea6555b0a6d.png">
